File: by-kotlin/kt-boj/solutions/11724-연결 요소의 개수.kt
import java.io.BufferedReader
import java.io.InputStreamReader
import java.util.*
import kotlin.collections.ArrayList

private lateinit var br: BufferedReader
private lateinit var token: StringTokenizer

fun main() {
    br = BufferedReader(InputStreamReader(System.`in`))
    token = StringTokenizer(br.readLine())
    val n = token.nextToken().toInt()
    val m = token.nextToken().toInt()
    val adj = arrayListOf<ArrayList<Int>>()
    val visited = arrayListOf<Boolean>()
    var result = 0

    for (i in 0..n) {
        adj.add(arrayListOf())
        visited.add(false)
    }

    for (i in 0 until m) {
        token = StringTokenizer(br.readLine())
        val x = token.nextToken().toInt()
        val y = token.nextToken().toInt()
        adj[x].add(y)
        adj[y].add(x)
    }

    (1..n).forEach { result += dfs(visited, adj, it) }

    print(result)
}

fun dfs(visited: ArrayList<Boolean>, adj: ArrayList<ArrayList<Int>>, v: Int): Int {
    if (visited[v]) {
        return 0
    }

    visited[v] = true
    for (k in adj[v]) {
        if (!visited[k]) {
            dfs(visited, adj, k)
        }
    }
    return 1
}

File: by-python/boj/11724-연결 요소의 개수.py
import sys
sys.setrecursionlimit(10000000)

n, m = map(int, sys.stdin.readline().split())
adj = [[] for _ in range(n + 1)]
visited = [False] * (n + 1)


def dfs(v):
    if visited[v]:
        return 0

    visited[v] = True
    for k in adj[v]:
        if not visited[k]:
            dfs(k)
    return 1


result = 0
for _ in range(m):
    x, y = map(int, sys.stdin.readline().split())
    adj[x].append(y)
    adj[y].append(x)

for i in range(1, n + 1):
    result += dfs(i)

print(result)
